src/fft128/f128_impl.rs

```diff
@@ -0,0 +1,1075 @@
+        };
+
+        if p == 0 {
+            (s, c)
+        } else if p == 1 {
+            (c, -s)
+        } else if p == -1 {
+            (-c, s)
+        } else {
+            (-s, -c)
+        }
+    }
+}
+
+#[allow(clippy::approx_constant)]
+impl f128 {
+    pub const PI: Self = f128(3.141592653589793, 1.2246467991473532e-16);
+
+    const SINPI_TAYLOR: &'static [Self; 9] = &[
+        f128(-5.16771278004997, 2.2665622825789447e-16),
+        f128(2.5501640398773455, -7.931006345326556e-17),
+        f128(-0.5992645293207921, 2.845026112698218e-17),
+        f128(0.08214588661112823, -3.847292805297656e-18),
+        f128(-0.0073704309457143504, -3.328281165603432e-19),
+        f128(0.00046630280576761255, 1.0704561733683463e-20),
+        f128(-2.1915353447830217e-5, 1.4648526682685598e-21),
+        f128(7.952054001475513e-7, 1.736540361519021e-23),
+        f128(-2.2948428997269873e-8, -7.376346207041088e-26),
+    ];
+
+    const COSPI_TAYLOR: &'static [Self; 9] = &[
+        f128(-4.934802200544679, -3.1326477543698557e-16),
+        f128(4.0587121264167685, -2.6602000824298645e-16),
+        f128(-1.3352627688545895, 3.1815237892149862e-18),
+        f128(0.2353306303588932, -1.2583065576724427e-18),
+        f128(-0.02580689139001406, 1.170191067939226e-18),
+        f128(0.0019295743094039231, -9.669517939986956e-20),
+        f128(-0.0001046381049248457, -2.421206183964864e-21),
+        f128(4.303069587032947e-6, -2.864010082936791e-22),
+        f128(-1.3878952462213771e-7, -7.479362090417238e-24),
+    ];
+
+    const SIN_K_PI_OVER_16_TABLE: &'static [Self; 4] = &[
+        f128(0.19509032201612828, -7.991079068461731e-18),
+        f128(0.3826834323650898, -1.0050772696461588e-17),
+        f128(0.5555702330196022, 4.709410940561677e-17),
+        f128(0.7071067811865476, -4.833646656726457e-17),
+    ];
+
+    const COS_K_PI_OVER_16_TABLE: &'static [Self; 4] = &[
+        f128(0.9807852804032304, 1.8546939997825006e-17),
+        f128(0.9238795325112867, 1.7645047084336677e-17),
+        f128(0.8314696123025452, 1.4073856984728024e-18),
+        f128(0.7071067811865476, -4.833646656726457e-17),
+    ];
+}
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64", doc))]
+#[cfg_attr(docsrs, doc(cfg(any(target_arch = "x86", target_arch = "x86_64"))))]
+pub mod x86 {
+    #[cfg(target_arch = "x86")]
+    use core::arch::x86::*;
+    #[cfg(target_arch = "x86_64")]
+    use core::arch::x86_64::*;
+
+    pulp::simd_type! {
+        /// Avx SIMD type.
+        pub struct Avx {
+            pub sse: "sse",
+            pub sse2: "sse2",
+            pub avx: "avx",
+            pub fma: "fma",
+        }
+    }
+
+    #[cfg(feature = "nightly")]
+    pulp::simd_type! {
+        /// Avx512 SIMD type.
+        pub struct Avx512 {
+            pub sse: "sse",
+            pub sse2: "sse2",
+            pub avx: "avx",
+            pub avx2: "avx2",
+            pub fma: "fma",
+            pub avx512f: "avx512f",
+        }
+    }
```

Comment:
doesn't pulp define these types itself ?

Reply:
`pulp` defines the just atomic simd units so it can just reexport the low level intrinsics. since the number of combinations would be too great to include them all

simd_type allows us to merge them to target a specific combination, since we usually want to use instructions from different subsets

Reply:
understood, thanks !